### ISSUE TITLE
Error: stat /Users/xxxxx/.kube/config: no such file or directory when trying to use with EKS

### ISSUE THREAD

**Issue by jbilliau5668** [hashicorp/terraform-provider-helm#418]
### Terraform Version
12.18

### Affected Resource(s)
- helm_release

### Terraform Configuration Files
```hcl
data "aws_eks_cluster_auth" "cluster-auth" {
  depends_on = [aws_eks_cluster.default]
  name       = "aws_eks_cluster.default"
}

provider "helm" {
  alias = "eks"
  kubernetes {
    # config_path = "${path.module}/.kube/config"
    host                   = aws_eks_cluster.default.endpoint
    cluster_ca_certificate = aws_eks_cluster.default.certificate_authority[0].data #base64decode(aws_eks_cluster.default.certificate_authority.0.data)
    token                  = data.aws_eks_cluster_auth.cluster-auth.token
    load_config_file       = false
  }
}

resource "helm_release" "controllers" {
  name  = "controllers"
  atomic = true
  chart = "./controllers"
  depends_on = [
    aws_eks_cluster.default,
    local_file.kubeconfig,
    aws_autoscaling_group.default
  ]

  set {
    name  = "clusterName"
    value = aws_eks_cluster.default.id
  }

  set {
    name  = "region"
    value = var.aws_region
  }

  set {
    name  = "zone"
    value = var.route53zone
```

### Expected Behavior
It creates a Helm release with my the YAML in my _controllers_ folder.

### Actual Behavior
Receive this errors instead:

_Error: stat /Users/xxxxx/.kube/config: no such file or directory

  on main.tf line 610, in resource "helm_release" "controllers":
 610: resource "helm_release" "controllers" {_

Here's what I know; if I have a ~/.kube/config file present, everything works just fine. Thus, when running **terraform apply** locally on my laptop, it works. However, I'm trying to replicate running terraform apply in a fresh container in our CD platform, which won't have that file.

Now, oddly enough, even though the rest of my EKS TF module has commands all over the place using _"${path.module}/.kube/config"_ as the source for the kubeconfig and work just fine, Helm3 provider does not. It INSISTS I have a config file at ~/.kube/config, even though I:

1. Have a perfectly useable one at  ${path.module}/.kube/config, path.module being the root of my module repo. This is useable with literally everything else, including various kubectl commands.
2. Tried, as you can see above, not even using a file and just configuring the provider with the cert, token, etc., as well as specifying load_config_file to false. The provider seems to not care about any of this, everything be damned if there is no ~/.kube.config.

### Steps to Reproduce
terraform apply against an EKS cluster.

**Comment by matthew-palmer** [comment]
@jbilliau5668 This has been noted from me as well. See #138 as this could be semi related to the way the kubernetes closure functions. https://github.com/terraform-providers/terraform-provider-helm/issues/138#issuecomment-580453704

The only way to prevent this is to expose the `load_config_file` from an environment variable. This was mentioned to be fixed in v1, but it may be around still. Try to source the value from `KUBE_LOAD_CONFIG_FILE` set to `"false"` and see if the behavior persists. This is by no means a fix, but it has worked for me to get to another bug with EKS.

**Comment by jbilliau5668** [comment]
So that got me past my stat error, but since I am no longer supplying a kubecong file, I needed to specify the authentication for the k8 cluster through the provider block explicitly, like this:

```
provider "helm" {
  kubernetes {
    host                   = aws_eks_cluster.default.endpoint
    cluster_ca_certificate = base64decode(aws_eks_cluster.default.certificate_authority.0.data)
    token                  = data.aws_eks_cluster_auth.cluster-auth.token
    load_config_file       = false
  }
}
```

Then I get my next error: "Kubernetes cluster unreachable". I highyl suspect this is because Helm is creating a kubeconfig file on the fly with the above variables and that wont work with EKS; it needs the aws-iam-authenticator to authenticate with the cluster.

So basically, I have 2 options:

1.) Pass in kubeconfig file that is EKS-specific and inject cluster and token name in it for Helm to use. That doesn't work since Helm tries to load at build time and I get the stat error.
2.) I configure variables to have Helm build a kubeconfig file dynamically. I get past the stat error but since the kubeconfig file isn't EKS-compatible, it blows up anyway.

Was this provider tested with EKS at all? It seems to be completely non-functional, i don't understand how anyone is using this.

**Comment by matthew-palmer** [comment]
The kubeconfig file doesn't need to be "EKS specific" as EKS still is just a kubernetes endpoint provided by Amazon. If the generated kubeconfig has all the valid information, it will connect. You could try running the entire process again with `TF_LOG=trace` so you get a bit more information on what is actually happening behind the scenes. 
The unfortunate truth is that for my work, the only way I was able to get the helm provider to even connect with EKS was to create a stopgap `kubeconfig` template generation module with the required values (host, CA cert, token etc.) and then write that to a file and use it as a source to load_config_file.
```
# kubeconfig.tpl
apiVersion: v1
kind: Config
clusters:
- name: "cluster"
  cluster:
    certificate-authority-data: "${cluster_auth_data}"
    server: "${cluster_url}"
    api-version: v1
users:
- name: "user"
  user:
    token: "${cluster_token}"
contexts:
- name: "eks"
  context:
    user: "user"
    cluster: "cluster"
current-context: "eks"
```

```
# kubeconfig.tf
locals {
  kube_template = "${path.module}/kubeconfig.tpl"
}

data "template_file" "kubeconfig" {
  template = "${file(local.kube_template)}"

  vars = {
    cluster_url   = data.aws_eks_cluster.this.endpoint
    cluster_token = data.aws_eks_cluster_auth.this.token
    cluster_auth_data = data.aws_eks_cluster.this.certificate_authority.0.data
  }
}

resource "local_file" "kubeconfig" {
  content  = data.template_file.kubeconfig.rendered
  filename = var.file_location
}
```

Also note that none of this will work if the AWS user executing this does NOT have access to the eks cluster. Best approach I have seen, which may be up for discussion, is to create and manage EKS clusters through an assumable role, and run this under said role.

Best way is to run this module, use the generated kubeconfig and test your own connection first locally. If this works, note the kubeconfig token defaults to ~10 minutes, and will time out after that.

**Comment by jbilliau5668** [comment]
tried this, doesn't work; the provider blows up with the same error:

```
_Error: stat /Users/xxxxx/.kube/config: no such file or directory

on main.tf line 610, in resource "helm_release" "controllers":
610: resource "helm_release" "controllers" {_
```

It seems to require that file to be present at build time, even though any localfile resource won't get created until apply time. I don't understand how this is working for anyone.

**Comment by matthew-palmer** [comment]
Can you verify the kubeconfig file functions after its generation? If so, you can also try setting the environment variable `KUBECONFIG` to the generated location, as it seems like it is still targeting the default path for kubeconfig. 
Terraform doesn't do well with sequential operations, so if generating the kubeconfig AND using it are in the same module, it would be best to change OOP to target the specific generation module, then run resources requiring said config.

**Comment by jbilliau5668** [comment]
sorry for the late response, been busy.

So yes the kubeconfig works after generation; we are currently installing multiple controls as well as Istio in our Terraform EKS module, just with null_resources running the helm binary. We'd like to convert to using the native helm provider, and thus, have ran into this issue.

Here is my kubeconfig template, a file that sits in our repo and is injected at deploy time:

```
apiVersion: v1
clusters:
- cluster:
    server: ${eks_cluster_endpoint}
    certificate-authority-data: ${eks_cluster_ca}
  name: kubernetes
contexts:
- context:
    cluster: kubernetes
    user: aws
  name: aws
current-context: aws
kind: Config
preferences: {}
users:
- name: aws
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1alpha1
      command: aws-iam-authenticator
      args:
        - "token"
        - "-i"
        - "${eks_cluster_name}"
```

Then, here is where I populate those variables:

```
resource "local_file" "kubeconfig" {
  content = templatefile("${path.module}/files/kubeconfig", {
    eks_cluster_endpoint = aws_eks_cluster.default.endpoint
    eks_cluster_ca       = aws_eks_cluster.default.certificate_authority[0].data
    eks_cluster_name     = "${local.eks_cluster_name}"
  })
  filename = "${path.module}/.kube/config"
}
```

And this is where I point the Helm provider to that file and try to install a Helm chart:

```
provider "helm" {
  kubernetes {
    config_path = "${path.module}/.kube/config"
  }
}

resource "helm_release" "splunk_logger" {
  count = var.enable_splunk_connect_logger ? 1 : 0
  name  = "splunk-connector-logs"
  depends_on = [
    aws_eks_cluster.default,
    null_resource.update_config_map_aws_auth,
    aws_autoscaling_group.default,
    local_file.kubeconfig
  ]
  repository = "https://git.rockfin.com/pages/BAnderson5/splunk-connect-for-kubernetes-releases/"
  chart      = "splunk-kubernetes-logging"
  values = [
    "${file(local.splunk_helm_chart_values_file)}"
  ]
  timeout = "300"

  set {
    name  = "global.splunk.hec.token"
    value = var.splunk_token
  }
  set {
    name  = "global.kubernetes.clusterName"
    value = local.eks_cluster_name
  }
  set {
    name  = "global.splunk.hec.indexName"
    value = local.splunk_index
  }
  set {
    name  = "k8sMetadata.podLabels"
    value = "{ ${join(",", var.pod_labels)} }"
  }
}
```

Problem is, the file isn't put at "${path.module}/.kube/config" until apply time, and Helm seems to want it at build time, which makes no sense.  I've also tried setting the file path via an env variable (KUBECONFIG), no dice.

**Comment by jbilliau5668** [comment]
Also, what did you mean @matthew-palmer that the kubeconfig isn't EKS-specific? As you can see in my example above, my user block is:

```
users:
- name: aws
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1alpha1
      command: aws-iam-authenticator
      args:
        - "token"
        - "-i"
        - "${eks_cluster_name}"
```

Note the command: aws-iam-authenticator, and then passing in the token arguments....how would helm know to do that when it builds a kubeconfig from the parameters you are passing in when you do this?

```
provider "helm" {
  kubernetes {
    host                   = aws_eks_cluster.default.endpoint
    cluster_ca_certificate = base64decode(aws_eks_cluster.default.certificate_authority.0.data)
    token                  = data.aws_eks_cluster_auth.cluster-auth.token
    load_config_file       = false
  }
}
```

I just figured that's why I was getting Kubernetes cluster unreachable, no?

**Comment by dak1n1** [comment]
Hi, everyone! Thanks for filing this issue. I'm hoping that we can do some troubleshooting and find a solution that will work for everyone.

I did some testing with this and (as a newer user myself) I did find it more difficult to use the helm provider with EKS, compared to using a local test cluster. This could indicate that we need to add some documentation to guide users through the process, as it took a bit of trial and error for me to get this working. Here's what ended up working for me:

```
[dakini@dax 20200417]$ cat main.tf 
provider "helm" {
  version = "1.1.1"
  kubernetes {
    load_config_file  = false
    host     = "https://A1D100D49B28DDE487179.yl4.us-west-2.eks.amazonaws.com"
    token    = file("/home/dakini/work/20200417/token")
    cluster_ca_certificate = file("/home/dakini/work/20200417/cluster-ca-cert.pem")
  }
}

data "helm_repository" "stable" {
  name = "stable"
  url  = "https://kubernetes-charts.storage.googleapis.com"
}

resource "helm_release" "consul" {
  name       = "myconsul"
  repository = "stable"
  chart      = "consul"
  version    = "3.9.5"
}
```

I removed ~/.kube and unset the $KUBECONFIG environment variable before running `terraform apply`. Thankfully, it didn't seem to require either of those to exist.

But I did have to take information that would normally be in my kubeconfig, such as the CA cert and server hostname, and insert those into my terraform config, as you can see above. (These files are not base64 encoded).

One part that was not intuitive for me was having to set up a service account on the cluster prior to running this, since helm seems to require fairly broad permissions.

Here's everything I did to set up the cluster prior to running the helm provider:

```
# build a test EKS cluster
cd terraform-provider-kubernetes/kubernetes/test-infra/eks
AWS_PROFILE=default AWS_REGION=us-west-2 time terraform apply -var 'kubernetes_version=1.15' --auto-approve

KUBECONFIG=$(terraform output kubeconfig_path)

kubectl create serviceaccount testsa
kubectl get serviceaccount testsa -o yaml |grep token # grab secret name

# cd into the directory with my helm main.tf file
cd /home/dakini/work/20200417

# set up the files that main.tf will read
kubectl get secret testsa-token-tqk4r --template={{.data.token}} |base64 -d > token
grep certificate-authority-data $KUBECONFIG | awk '{print $2}' |base64 -d > cluster-ca-cert.pem
grep server $KUBECONFIG | awk '{print $2}' # grab EKS endpoint name
vim main.tf # add host to this file manually

# add some pretty broad permissions for helm to apply the chart
kubectl create clusterrolebinding helm-admin --clusterrole=cluster-admin --serviceaccount=default:testsa

unset KUBECONFIG
rm -rf ~/.kube

terraform init
terraform apply
```

I ran this to verify that there really is no default kubeconfig it's reading. The errors are expected:

```
[dakini@dax 20200417]$ helm list
Error: Kubernetes cluster unreachable
[dakini@dax 20200417]$ kubectl get pods
The connection to the server localhost:8080 was refused - did you specify the right host or port?
```

And I verified that the resources did indeed get applied by terraform, so this looks good:

```
[dakini@dax 20200417]$ helm list --kubeconfig=/home/dakini/go/src/github.com/hashicorp/terraform-provider-kubernetes/kubernetes/test-infra/eks/kubeconfig_k8s-acc-f8fa
NAME            NAMESPACE       REVISION        UPDATED                                 STATUS          CHART           APP VERSION
myconsul        default         1               2020-04-17 12:17:56.132391701 -0700 PDT deployed        consul-3.9.5    1.5.3

[dakini@dax 20200417]$ kubectl get pods --kubeconfig=/home/dakini/go/src/github.com/hashicorp/terraform-provider-kubernetes/kubernetes/test-infra/eks/kubeconfig_k8s-acc-f8fa
NAME         READY   STATUS    RESTARTS   AGE
myconsul-0   1/1     Running   0          117s
myconsul-1   1/1     Running   0          93s
myconsul-2   1/1     Running   0          71s
```

I'm running these versions, FYI:
```
[dakini@dax 20200417]$ terraform version
Terraform v0.12.24
+ provider.helm v1.1.1
```
I also tested with these versions:
```
[dakini@dax 20200417]$ terraform0.12.18 version
Terraform v0.12.18
+ provider.helm v1.1.1
```

@jbilliau5668 Are you also using v1.1.1 of the provider? And are you running terraform against an EKS cluster that already exists (similar to my example), or is yours a single terraform run that creates the EKS cluster *and* applies the helm chart? Thanks!

**Comment by dak1n1** [comment]
It looks like there is also another approach to this, potentially. There's an exec block that was missing from our documentation until yesterday. https://github.com/terraform-providers/terraform-provider-helm/pull/469/files?short_path=12848d8#diff-12848d8d421c2fa28469f9cfc40602f0 

This would let you run `aws-iam-authenticator` instead of creating a service account like I was doing:

```
provider "helm" {
  kubernetes {
    load_config_file   = "false"
    host                     = "<EKS endpoint>"
    exec  {
      api_version = "client.authentication.k8s.io/v1alpha1"
      command= "aws-iam-authenticator"
      args=[ "token", "-i", "<EKS cluster name>"]
    }
  }
}
```

**Comment by codedumper** [comment]
I use another way: 
Output from my eks modules: 
```
data "aws_eks_cluster" "cluster" {
  name = module.eks.cluster_id
}

data "aws_eks_cluster_auth" "cluster" {
  name = module.eks.cluster_id
}

output "cluster_certificate" {
  description = "Certificate for EKS control plane."
  value       = base64decode(data.aws_eks_cluster.cluster.certificate_authority.0.data)
}

output "cluster_endpoint" {
  description = "Endpoint for EKS control plane."
  value       = module.eks.cluster_endpoint
}

output "cluster_token" {
  description = "Token for EKS control plane."
  value       = data.aws_eks_cluster_auth.cluster.token
}
```
here is my main modules call:
```
module "k8s-cluster" {
  source                  = "./modules/k8s-cluster"
  ...
}

module "nginx-ingress-controller" {
  source              = "./modules/nginx-ingress-controller"
  cluster_name        = var.cluster_name
  dns_zone_name       = var.dns_zone_name
  cluster_certificate = module.k8s-cluster.cluster_certificate
  cluster_endpoint    = module.k8s-cluster.cluster_endpoint
  cluster_token       = module.k8s-cluster.cluster_token
}
```
Get it in my next modules with: 
```
provider "helm" {
  version = "~> 1.1.1"
  kubernetes {
    load_config_file       = false
    host                   = var.cluster_endpoint
    cluster_ca_certificate = var.cluster_certificate
    token                  = var.cluster_token
  }
}
```

**Comment by dak1n1** [comment]
Great, thanks for that example. I'll leave this open for a little while to see if anyone is still getting the original error.

**Comment by mcordovaf** [comment]
I'm facing the same issue, Is there any way to male this work?

Thanks!! 😊

**Comment by dak1n1** [comment]
In this comment, I posted the steps that got it working for me. https://github.com/terraform-providers/terraform-provider-helm/issues/418#issuecomment-615438412

I wasn't able to reproduce the original issue. It worked for me. Please post a reproducer if the issue is ongoing. That will help us to see what's going wrong. Thanks!

Another option is using the exec block: https://www.terraform.io/docs/providers/kubernetes/index.html#exec

**Comment by aareet** [comment]
Closing since the issue has waited 15 days for a response.
